File: Task1/round_robin.py
"""
ST5004CEM Operating Systems and Security
Task 1 - Round Robin Scheduling Simulation

Author: Trisha Rajbanshi
"""

from collections import deque
import time


class Process:
    def __init__(self, pid, burst_time):
        self.pid = pid
        self.burst_time = burst_time


def round_robin(processes, quantum):
    queue = deque(processes)

    print("=" * 60)
    print("ROUND ROBIN CPU SCHEDULER")
    print("=" * 60)

    while queue:

        process = queue.popleft()

        execution_time = min(quantum, process.burst_time)

        print(f"\nRunning {process.pid}")
        print(f"CPU Time Used : {execution_time}")

        time.sleep(1)

        process.burst_time -= execution_time

        if process.burst_time > 0:
            print(f"Remaining Burst Time : {process.burst_time}")
            queue.append(process)

        else:
            print(f"{process.pid} Completed Successfully.")

    print("\nAll Processes Completed.")
    print("=" * 60)


def main():

    processes = [

        Process("P1", 5),

        Process("P2", 3),

        Process("P3", 7)

    ]

    quantum = 2

    round_robin(processes, quantum)


if __name__ == "__main__":
    main()

File: Task1/round_robin.c
#include <stdio.h>

int main()
{
    int n, tq;
    int burst[20], remaining[20];
    int waiting[20], turnaround[20];
    int completed = 0;
    int time = 0;

    printf("===== Round Robin CPU Scheduling =====\n\n");

    printf("Enter number of processes: ");
    scanf("%d", &n);

    printf("Enter Time Quantum: ");
    scanf("%d", &tq);

    for(int i = 0; i < n; i++)
    {
        printf("Burst Time for Process P%d: ", i + 1);
        scanf("%d", &burst[i]);

        remaining[i] = burst[i];
    }

    while(completed < n)
    {
        for(int i = 0; i < n; i++)
        {
            if(remaining[i] > 0)
            {
                if(remaining[i] <= tq)
                {
                    time += remaining[i];
                    remaining[i] = 0;

                    turnaround[i] = time;
                    waiting[i] = turnaround[i] - burst[i];

                    completed++;
                }
                else
                {
                    remaining[i] -= tq;
                    time += tq;
                }
            }
        }
    }

    printf("\n-------------------------------------------------\n");
    printf("Process\tBurst Time\tWaiting Time\tTurnaround Time\n");
    printf("-------------------------------------------------\n");

    float totalWaiting = 0;
    float totalTurnaround = 0;

    for(int i = 0; i < n; i++)
    {
        printf("P%d\t%d\t\t%d\t\t%d\n",
               i + 1,
               burst[i],
               waiting[i],
               turnaround[i]);

        totalWaiting += waiting[i];
        totalTurnaround += turnaround[i];
    }

    printf("-------------------------------------------------\n");

    printf("\nAverage Waiting Time    : %.2f", totalWaiting / n);
    printf("\nAverage Turnaround Time : %.2f\n", totalTurnaround / n);

    return 0;
}
